### DISCUSSION TITLE
Does Pyboard support 9-bit UART?

### DISCUSSION THREAD

**Discussion by daleka** [micropython/micropython#16152] [STM32 / Pyboard] [answered]
Hi everyone!

I’m working on a project that requires 9-bit UART support for data transmission. Does the standard MicroPython firmware for Pyboard support a 9-bit UART mode? If not, are there any custom builds available with this functionality?

I’ve seen that the STM32 port had modifications allowing 9-bit mode, but I’m not sure if this applies to Pyboard specifically. Also, are there any plans to add full support for 9-bit UART on Pyboard in future MicroPython firmware releases?

Any information or advice would be greatly appreciated!

**Comment by robert-hh** [comment]
The 9-bit mode should be available at the PyBoard and at the other STM32 devices. As far as I can tell from the code, the only condition is that parity has to be None.

**Comment by daleka** [comment]
Hi Robert,

Thank you for the quick response! I’ve set up UART with 9-bit mode, but I’m encountering an issue when data is received. Specifically, as soon as data arrives, I get the following error:
Traceback (most recent call last):
  File "<stdin>", line 19, in <module>
  File "<stdin>", line 14, in read_uart
OSError: [Errno 5] EIO

If I switch the UART to 8-bit mode (bits=8), everything works fine, and data is received without any issues. Here are the UART settings I’m using:
uart = UART(4, baudrate=9600, bits=9, parity=None, stop=1)

For reference, here’s the version of MicroPython I’m using:
(sysname='pyboard', nodename='pyboard', release='1.24.0', version='v1.24.0 on 2024-10-25', machine='PYBv1.1 with STM32F405RG')

Are there any known limitations or additional configurations required for 9-bit mode to work correctly on PyBoard? Or could this be an issue with my specific version of MicroPython?

Thanks again for your help, and I would appreciate any insights or suggestions!

**Reply by robert-hh** [reply]
In 9 bit mode each 9 bit data item is packed into 2 bytes, low order byte first.  So you have to take care to send and receive even number of bytes.

**Reply by daleka** [reply]
My task involves implementing the MDB protocol on a microcontroller as a slave device connected to an MDB (Multi-Drop Bus) system, which is widely used in vending machines. The MDB protocol operates in a master-slave configuration, where the vending machine controller (VMC) acts as the master, and all connected devices, including mine, are slaves.

MDB uses a 9-bit format for data transmission: 1 start bit, 8 data bits, and an additional 9th bit, which indicates whether the byte is an address or data byte. Each message begins with an address byte, where this additional bit is set, and all following bytes are transmitted as data. This allows the master to distinguish between address and data bytes.

In my case, as a slave device, I only need to correctly receive data without controlling the transmission process and identify whether each received byte is an address or data byte. My implementation involves reading each byte, checking the 9th bit, and handling it accordingly. The issue is that I encounter the [Errno 5] EIO error specifically when data is received in 9-bit mode.

If MicroPython cannot fully support 9-bit mode to handle MDB correctly, I may need to consider alternative solutions or even rethink using MicroPython for this project.

I would appreciate any recommendations on configuration or confirmation if there are technical limitations preventing the implementation of this protocol in MicroPython.

**Reply by robert-hh** [reply]
Instead of reading one byte, you have to read 2 bytes. So call uart.read(2) to get one 9 bit value, uart.read(4) to get two values or any other even number of bytes. You can check uart.any() to test, how many 9 bit items are available. If uart.any() returns 1, call uart.read(2), if it returns 2, call uart.read(4), ...... Of any byte pair, the second one contains the 9th bit. If for instance you wrote msg = read(4), then msg[1] has the 9th bit of the first word, msg[3] has the 9th bit of the second word. Pretty straight!
If you call uart.read() with an odd number as argument, you'll get the EOI error.

**Reply by daleka** [reply]
Thank you for your guidance! Following your suggestion to read two bytes for each 9-bit value solved the issue. Everything now works perfectly, and my setup correctly distinguishes address and data bytes without errors

I really appreciate your help!

**Reply by daleka** [reply]
> Instead of reading one byte, you have to read 2 bytes. So call uart.read(2) to get one 9 bit value, uart.read(4) to get two values or any other even number of bytes. You can check uart.any() to test, how many 9 bit items are available. If uart.any() returns 1, call uart.read(2), if it returns 2, call uart.read(4), ...... Of any byte pair, the second one contains the 9th bit. If for instance you wrote msg = read(4), then msg[1] has the 9th bit of the first word, msg[3] has the 9th bit of the second word. Pretty straight! If you call uart.read() with an odd number as argument, you'll get the EOI error.

"Thank you again for helping me with receiving 9-bit data! The solution works perfectly, and I can now accurately distinguish between address and data bytes.

However, I’m facing an issue with sending 9-bit data. The peripheral device on the MDB protocol is not using MicroPython and seems to expect each byte as a single 9-bit packet. Currently, I’m sending it as 8 bits plus an additional bit, which likely results in the device not responding to my messages.

Are there any recommendations on how to send a full 9-bit byte in a single transmission cycle? I would greatly appreciate any advice on overcoming this limitation in MicroPython."

**Reply by robert-hh** [reply, marked as answer]
Same as with read. You have to pack a 9 bit value into 2 bytes, low order bits first. And then use uart.write() with the even number of bytes for sending. e.g. to send the value 0x161 as 9 bit quantity, use uart.write("\x61\x01"). You may create a separate function for format the message. It is important that you send the whole message in a single call to uart.write().
